### PR TITLE
This should fix MySQL 5.7 issues caused by ONLY_FULL_GROUP_BY

### PR DESCRIPTION
**Pull request by wucdbm**
First of all, I have to say that this is my first pull request ever created

This should fix MySQL 5.7 issues caused by ONLY_FULL_GROUP_BY

The proposed code works for me, maybe others could share opinions on a fix for this, it is very annoying when working with new versions of MySQL

### CODE REVIEW

**Comment by wucdbm** [comment]
#5622

**Comment by buffcode** [comment]
CI currently fails because of tests expect string equality. PR will probably need to contain the updated tests in order to get merged.

**Comment by wucdbm** [comment]
I am aware. 

However, I lack the time to dig into it and change tests accordingly. I also have no clue if and if so, how this would affect other DBMS, so I'll simply leave my code here and hope it would help you guys get a decent solution out of it :)

File: lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php [hunk 1]
```diff
@@ -348,33 +348,56 @@ private function addMissingItemsFromOrderByToSelect(SelectStatement $AST)
...
         // statement
-        if (! $orderByClause instanceof OrderByClause) {
+        if (!$orderByClause instanceof OrderByClause) {
             return $sql;
         }
 
         // Rebuild the order by clause to work in the scope of the new select statement
         /* @var array $orderBy an array of rebuilt order by items */
         $orderBy = $this->rebuildOrderByClauseForOuterScope($orderByClause);
+        $orderByFields = str_replace([' DESC', ' ASC'], ['', ''], $orderBy);
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php]: You may have something like `ORDER BY ASCII ASC` - that would be replaced with `ORDER BYII` here - no-go.

File: lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php [hunk 2]
```diff
@@ -348,33 +348,56 @@ private function addMissingItemsFromOrderByToSelect(SelectStatement $AST)
...
+                unset($orderByFields[$k]);
+            }
+        }
+
+        foreach ($sqlIdentifier as $k => $v) {
+            $innerSqlIdentifier[$k] = 'dctrn_result_inner.' . $v;
+            $sqlIdentifier[$k] = 'dctrn_result.' . $v;
+        }
+
+        // add the
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php]: comment incomplete?

File: lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php [hunk 3]
```diff
@@ -348,33 +348,56 @@ private function addMissingItemsFromOrderByToSelect(SelectStatement $AST)
...
         }
 
         // Rebuild the order by clause to work in the scope of the new select statement
         /* @var array $orderBy an array of rebuilt order by items */
         $orderBy = $this->rebuildOrderByClauseForOuterScope($orderByClause);
+        $orderByFields = str_replace([' DESC', ' ASC'], ['', ''], $orderBy);
+
+        $innerSqlIdentifier = [];
+
+        foreach ($orderByFields as $k => $v) {
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php]: `$k` and `$v` are not good names

File: lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php [hunk 4]
```diff
@@ -348,33 +348,56 @@ private function addMissingItemsFromOrderByToSelect(SelectStatement $AST)
...
+            // remove fields that are selected by identifiers,
+            // if those are ordered by in the query
+            if (in_array($v, $sqlIdentifier)) {
+                unset($orderByFields[$k]);
+            }
+        }
+
+        foreach ($sqlIdentifier as $k => $v) {
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php]: Same here

File: lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php [hunk 5]
```diff
@@ -348,33 +348,56 @@ private function addMissingItemsFromOrderByToSelect(SelectStatement $AST)
...
+            $innerSqlIdentifier[$k] = 'dctrn_result_inner.' . $v;
+            $sqlIdentifier[$k] = 'dctrn_result.' . $v;
+        }
+
+        // add the
+        foreach ($orderByFields as $k => $v) {
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php]: Same here: `$k`, `$v` are not good names

File: lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php [hunk 6]
```diff
@@ -348,33 +348,56 @@ private function addMissingItemsFromOrderByToSelect(SelectStatement $AST)
...
-            'SELECT DISTINCT %s FROM (%s) dctrn_result ORDER BY %s',
-            implode(', ', $sqlIdentifier),
+            'SELECT DISTINCT %s FROM (%s) dctrn_result_inner ORDER BY %s',
+            implode(', ', $innerSqlIdentifier),
             $innerSql,
             implode(', ', $orderBy)
         );
 
+        // now only select distinct identifier
+        $sql = sprintf('SELECT DISTINCT %s FROM (%s) dctrn_result', implode(', ', $sqlIdentifier), $sql);
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php]: Return directly

File: lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php (hunk 1 above)
Comment (Ocramius) [inline lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php]: This calls for additional test scenarios for both descending and ascending sorting clauses

File: lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php [hunk 7]
```diff
@@ -348,33 +348,56 @@ private function addMissingItemsFromOrderByToSelect(SelectStatement $AST)
...
      *
      * @return string
      */
-    private function preserveSqlOrdering(array $sqlIdentifier, $innerSql, $sql, $orderByClause)
-    {
+    private function preserveSqlOrdering(array $sqlIdentifier, $innerSql, $sql, $orderByClause) {
         // If the sql statement has an order by clause, we need to wrap it in a new select distinct
         // statement
-        if (! $orderByClause instanceof OrderByClause) {
+        if (!$orderByClause instanceof OrderByClause) {
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php]: Please revert

File: lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php [hunk 8]
```diff
@@ -348,33 +348,56 @@ private function addMissingItemsFromOrderByToSelect(SelectStatement $AST)
     /**
      * Generates new SQL for statements with an order by clause
      *
-     * @param array           $sqlIdentifier
-     * @param string          $innerSql
-     * @param string          $sql
-     * @param OrderByClause   $orderByClause
+     * @param array $sqlIdentifier
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php]: please revert

File: lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php [hunk 9]
```diff
@@ -348,33 +348,56 @@ private function addMissingItemsFromOrderByToSelect(SelectStatement $AST)
...
+     * @param string $innerSql
+     * @param string $sql
+     * @param OrderByClause $orderByClause
      *
      * @return string
      */
-    private function preserveSqlOrdering(array $sqlIdentifier, $innerSql, $sql, $orderByClause)
-    {
+    private function preserveSqlOrdering(array $sqlIdentifier, $innerSql, $sql, $orderByClause) {
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php]: Please revert

File: lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php (hunk 1 above)
Comment (Ocramius) [inline lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryOutputWalker.php]: You probably want to match a series of [`((\S+)\s+(ASC|DESC)\s*,?)*`](https://regexper.com/#((%5CS%2B)%5Cs%2B(ASC%7CDESC)%5Cs*%2C%3F)*) here (or similar)

**Comment by netiul** [comment]
@wucdbm Any chance you get to implement the comments any time soon? It would be great to see this bug fixed.

**Comment by Shandur** [comment]
I ran into the same problem. The problem is still not fixed? @Ocramius

**Comment by theofidry** [comment]
@Shandur doesn't look like it. Maybe you can finish that PR? it doesn't look like much is missing from it.

**Comment by Shandur** [comment]
@theofidry I'll try to do it in few days.

**Comment by lukasneubauer** [comment]
hello guys, any progress with this?

**Comment by Shandur** [comment]
@lukasneubauer Hey. May was really busy for me, I didn't have any time to work with it. Gonna try to finish it the next week, I'll have some free time. I'll keep you posted.

**Comment by bourdeau** [comment]
Hello,

We are all waiting for your PR in our company and I guess we are not the only ones :)

**Comment by Shandur** [comment]
@bourdeau Hey. In order not to stop your developing, you can use this crap as temporary solution before executing a query.

`
private function disableMysqlMode()
    {
        if ($this->_em->getConnection()->getDriver()->getName() === 'pdo_mysql') {
            $sql = " SET sql_mode=(SELECT REPLACE(@@sql_mode, 'ONLY_FULL_GROUP_BY', '')) ";
            $this->_em->getConnection()->prepare($sql)->execute();
        }
    }
`

**Comment by theofidry** [comment]
@bourdeau then maybe on waiting on it someone could finish this up ;)

**Comment by lcobucci** [comment]
@Ocramius I think we can close this one and leave it to merge #6143 (seems more complete), agreed?

**Comment by Ocramius** [comment]
@lcobucci yeap.

**Comment by Ocramius** [comment]
Handled in https://github.com/doctrine/doctrine2/pull/6143

**Pull request closed without merging**